### ISSUE TITLE
[YSQL] Command timeouts and Tserver crashes when creating a foreign key

### ISSUE THREAD

**Issue by andrei-mart** [yugabyte/yugabyte-db#15220]
Jira Link: [DB-4399](https://yugabyte.atlassian.net/browse/DB-4399)
### Description

Reported by a customer, they tried to clone tables using `CREATE TABLE AS` statement:
```
create table mytest130 as select * from mytest129;
SELECT 2180302
alter table mytest130 add primary key (mykey asc);
ALTER TABLE
create table mychild130 as select * from mychild129;
SELECT 4360604
alter table mychild130 add primary key(mykey asc, myseq asc);
ALTER TABLE
alter table mychild130 add foreign key (mykey) references mytest129(mykey) on delete cascade;
ysqlsh:clone_tables.sql:5: ERROR: Timed out: Timed out waiting for Alter Table
create table mytest131 as select * from mytest129;
```
I ran into same or similar issue while trying to setup test environment, so problem is narrowed down to `ALTER TABLE ADD FOREIGN KEY` on a table with data.

I created a new rf=3 cluster:
```
$ bin/yb-ctl create --rf=3
Creating cluster.
Waiting for cluster to be ready.
----------------------------------------------------------------------------------------------------
| Node Count: 3 | Replication Factor: 3                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh                                                                 |
| YCQL Shell          : bin/ycqlsh                                                                 |
| YEDIS Shell         : bin/redis-cli                                                              |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /home/amartsin/yugabyte-data                                               |
----------------------------------------------------------------------------------------------------

For more info, please use: yb-ctl status
```
In the ysqlsh:
```
yugabyte=# create table mytest129(mykey int primary key, myvalue text);
CREATE TABLE
Time: 1163.959 ms (00:01.164)
yugabyte=# insert into mytest129 select i, 'mytest ' || i::text from generate_series(1, 2180302) i;
INSERT 0 2180302
Time: 242104.873 ms (04:02.105)
yugabyte=# create table mychild129(mykey int, myseq int, myvalue text, primary key (mykey, myseq));
CREATE TABLE
Time: 446.205 ms
yugabyte=# insert into mychild129 select i, j, 'mychild ' || i::text || ', ' || j::text  from generate_series(1, 2180302) i, generate_series(1, 2) j;
INSERT 0 4360604
Time: 608915.706 ms (10:08.916)
yugabyte=# alter table mychild129 add foreign key (mykey) references mytest129(mykey) on delete cascade;
ERROR:  Timed out waiting for Alter Table
Time: 60018.922 ms (01:00.019)
```
Subsequent command failed due to apparent crash of all the tservers:
```
yugabyte=# truncate mychild129;
FATAL:  terminating connection due to administrator command
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
Time: 0.507 ms
!> \q
$ bin/ysqlsh
ysqlsh: could not connect to server: Connection refused
	Is the server running on host "localhost" (127.0.0.1) and accepting
	TCP/IP connections on port 5433?
$ bin/yb-ctl status
----------------------------------------------------------------------------------------------------
| Node Count: 3 | Replication Factor: 3                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh                                                                 |
| YCQL Shell          : bin/ycqlsh                                                                 |
| YEDIS Shell         : bin/redis-cli                                                              |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /home/amartsin/yugabyte-data                                               |
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
| Node 1: yb-tserver (Stopped), yb-master (pid 39444)                                              |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh                                                                 |
| YCQL Shell          : bin/ycqlsh                                                                 |
| YEDIS Shell         : bin/redis-cli                                                              |
| data-dir[0]         : /home/amartsin/yugabyte-data/node-1/disk-1/yb-data                         |
| yb-tserver Logs     : /home/amartsin/yugabyte-data/node-1/disk-1/yb-data/tserver/logs            |
| yb-master Logs      : /home/amartsin/yugabyte-data/node-1/disk-1/yb-data/master/logs             |
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
| Node 2: yb-tserver (Stopped), yb-master (pid 39449)                                              |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.2:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh -h 127.0.0.2                                                    |
| YCQL Shell          : bin/ycqlsh 127.0.0.2                                                       |
| YEDIS Shell         : bin/redis-cli -h 127.0.0.2                                                 |
| data-dir[0]         : /home/amartsin/yugabyte-data/node-2/disk-1/yb-data                         |
| yb-tserver Logs     : /home/amartsin/yugabyte-data/node-2/disk-1/yb-data/tserver/logs            |
| yb-master Logs      : /home/amartsin/yugabyte-data/node-2/disk-1/yb-data/master/logs             |
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
| Node 3: yb-tserver (Stopped), yb-master (pid 39453)                                              |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.3:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh -h 127.0.0.3                                                    |
| YCQL Shell          : bin/ycqlsh 127.0.0.3                                                       |
| YEDIS Shell         : bin/redis-cli -h 127.0.0.3                                                 |
| data-dir[0]         : /home/amartsin/yugabyte-data/node-3/disk-1/yb-data                         |
| yb-tserver Logs     : /home/amartsin/yugabyte-data/node-3/disk-1/yb-data/tserver/logs            |
| yb-master Logs      : /home/amartsin/yugabyte-data/node-3/disk-1/yb-data/master/logs             |
----------------------------------------------------------------------------------------------------
```
After starting the server I was able to truncate the table create FK and load data successfully.
```
Starting cluster with base directory /home/amartsin/yugabyte-data
Waiting for cluster to be ready.
----------------------------------------------------------------------------------------------------
| Node Count: 3 | Replication Factor: 3                                                            |
----------------------------------------------------------------------------------------------------
| JDBC                : jdbc:postgresql://127.0.0.1:5433/yugabyte                                  |
| YSQL Shell          : bin/ysqlsh                                                                 |
| YCQL Shell          : bin/ycqlsh                                                                 |
| YEDIS Shell         : bin/redis-cli                                                              |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /home/amartsin/yugabyte-data                                               |
----------------------------------------------------------------------------------------------------

For more info, please use: yb-ctl status
$ bin/ysqlsh
Timing is on.
ysqlsh (11.2-YB-2.17.1.0-b0)
Type "help" for help.

yugabyte=# truncate mychild129;
TRUNCATE TABLE
Time: 210.391 ms
yugabyte=# alter table mychild129 add foreign key (mykey) references mytest129(mykey) on delete cascade;
ALTER TABLE
Time: 1270.232 ms (00:01.270)
yugabyte=# insert into mychild129 select i, j, 'mychild ' || i::text || ', ' || j::text  from generate_series(1, 2180302) i, generate_series(1, 2) j;
INSERT 0 4360604
Time: 1065774.318 ms (17:45.774)
```

**Comment by bmatican** [comment]
@andrei-mart When you see server crashes, can you also pick up the cause of the crash? Generally, it would be one of the following 3
1. FATAL files -- can check the respective tserver log dirs
1. core files -- path where these show up depends on your core pattern, but likely it's the dir where you started yb-ctl from
1. OOM -- unlikely in a local test

**Comment by andrei-mart** [comment]
There are FATALs.
Full logs:
[ghi-15220-logs.tar.gz](https://github.com/yugabyte/yugabyte-db/files/10157507/ghi-15220-logs.tar.gz)

**Comment by m-iancu** [comment]
```
F20221205 10:55:05 ../../src/yb/tablet/operations/operation_driver.cc:398] T ec08bcbdf9cf40c9ac58b339c0695c47 P 68ba367655f74cafb63c66c1fbed26ea S RD-P Ts { days: 19331 time: 10:54:05.448647 } kChangeMetadata (0x000055cdce798ab0): Apply failed: Timed out (yb/util/operation_counter.cc:166): Timed out waiting for all pending operations to complete. 1 transactions pending. Waited for 60.001s, request: tablet_id: "ec08bcbdf9cf40c9ac58b339c0695c47" schema { columns { id: 10 name: "mykey" type { main: INT32 } is_key: 1 is_hash_key: 1 is_nullable: 0 is_static: 0 is_counter: 0 sorting_type: 0 order: 1 pg_type_oid: 23 } columns { id: 11 name: "myseq" type { main: INT32 } is_key: 1 is_nullable: 0 is_static: 0 is_counter: 0 sorting_type: 3 order: 2 pg_type_oid: 23 } columns { id: 12 name: "myvalue" type { main: STRING } is_key: 0 is_nullable: 1 is_static: 0 is_counter: 0 sorting_type: 0 order: 3 pg_type_oid: 25 } table_properties { contain_counters: 0 is_transactional: 1 consistency_level: STRONG use_mangled_column_name: 0 is_ysql_catalog_table: 0 retain_delete_markers: 0 partitioning_version: 1 } colocated_table_id { } pgschema_name: "public" } schema_version: 1 new_table_name: "mychild129" dest_uuid: "68ba367655f74cafb63c66c1fbed26ea" propagated_hybrid_time: 6841411360544137216 alter_table_id: "000033e8000030008000000000004005" transaction_id: "d73347a6-c533-46de-bcc8-91f8112ae97f" should_abort_active_txns: 1
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/util/logging.cc:458:     @     0x7f6ad6300890  yb::LogFatalHandlerSink::send(int, char const*, char const*, int, tm const*, char const*, unsigned long)
    @     0x7f6ad53202c9 
    @     0x7f6ad5320d2f 
    @     0x7f6ad5323608 
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/tablet/operations/operation_driver.cc:398:     @     0x7f6addd2da4a  yb::tablet::OperationDriver::ApplyTask(long, std::vector<yb::OpId, std::allocator<yb::OpId>>*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/tablet/operations/operation_driver.cc:355:     @     0x7f6addd2d47f  yb::tablet::OperationDriver::ReplicationFinished(yb::Status const&, long, std::vector<yb::OpId, std::allocator<yb::OpId>>*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/consensus_round.cc:47:     @     0x7f6add675364  yb::consensus::ConsensusRound::NotifyReplicationFinished(yb::Status const&, long, std::vector<yb::OpId, std::allocator<yb::OpId>>*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/replica_state.cc:1393:     @     0x7f6add707bc0  yb::consensus::ReplicaState::NotifyReplicationFinishedUnlocked(scoped_refptr<yb::consensus::ConsensusRound> const&, yb::Status const&, long, std::vector<yb::OpId, std::allocator<yb::OpId>>*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/replica_state.cc:957:     @     0x7f6add70caab  yb::consensus::ReplicaState::ApplyPendingOperationsUnlocked(yb::OpId const&, yb::StronglyTypedBool<yb::consensus::CouldStop_Tag>)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/replica_state.cc:870:     @     0x7f6add70b8f2  yb::consensus::ReplicaState::AdvanceCommittedOpIdUnlocked(yb::OpId const&, yb::StronglyTypedBool<yb::consensus::CouldStop_Tag>)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/replica_state.cc:769:     @     0x7f6add70a9de  yb::consensus::ReplicaState::UpdateMajorityReplicatedUnlocked(yb::OpId const&, yb::OpId*, bool*, yb::OpId*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/raft_consensus.cc:1378:     @     0x7f6add6bb467  yb::consensus::RaftConsensus::UpdateMajorityReplicated(yb::consensus::MajorityReplicatedData const&, yb::OpId*, yb::OpId*)
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/consensus/consensus_queue.cc:1596:     @     0x7f6add64557a  yb::consensus::PeerMessageQueue::NotifyObserversOfMajorityReplOpChangeTask(yb::consensus::MajorityReplicatedData const&)
../../src/yb/gutil/bind_internal.h:204:                                                                 @     0x7f6add67304d  yb::internal::RunnableAdapter<void (yb::consensus::PeerMessageQueue::*)(yb::consensus::MajorityReplicatedData const&)>::Run(yb::consensus::PeerMessageQueue*, yb::consensus::MajorityReplicatedData const&)
../../src/yb/gutil/bind_internal.h:901:                                                                 @     0x7f6add672fb0  yb::internal::InvokeHelper<false, void, yb::internal::RunnableAdapter<void (yb::consensus::PeerMessageQueue::*)(yb::consensus::MajorityReplicatedData const&)>, void (yb::consensus::PeerMessageQueue*, yb::consensus::MajorityReplicatedData const&)>::MakeItSo(yb::internal::RunnableAdapter<void (yb::consensus::PeerMessageQueue::*)(yb::consensus::MajorityReplicatedData const&)>, yb::consensus::PeerMessageQueue*, yb::consensus::MajorityReplicatedData const&)
../../src/yb/gutil/bind_internal.h:1159:                                                                @     0x7f6add672f4e  yb::internal::Invoker<2, yb::internal::BindState<yb::internal::RunnableAdapter<void (yb::consensus::PeerMessageQueue::*)(yb::consensus::MajorityReplicatedData const&)>, void (yb::consensus::PeerMessageQueue*, yb::consensus::MajorityReplicatedData const&), void (yb::internal::UnretainedWrapper<yb::consensus::PeerMessageQueue>, yb::consensus::MajorityReplicatedData)>, void (yb::consensus::PeerMessageQueue*, yb::consensus::MajorityReplicatedData const&)>::Run(yb::internal::BindStateBase*)
../../src/yb/gutil/callback.h:411:                                                                      @     0x7f6ade5efeed  yb::Callback<void ()>::Run() const
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:343:     @     0x7f6ad6480b25  decltype(std::declval<yb::Callback<void ()>&>().*std::declval<void (yb::Callback<void ()>::*&)() const>()()) std::__invoke[abi:v15002]<void (yb::Callback<void ()>::*&)() const, yb::Callback<void ()>&, void>(void (yb::Callback<void ()>::*&)() const, yb::Callback<void ()>&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/bind.h:263:     @     0x7f6ad6480aac  std::__bind_return<void (yb::Callback<void ()>::*)() const, tuple<yb::Callback<void ()>>, tuple<>, __is_valid_bind_return<void (yb::Callback<void ()>::*)() const, tuple<yb::Callback<void ()>>, tuple<>>::value>::type std::__apply_functor[abi:v15002]<void (yb::Callback<void ()>::*)() const, tuple<yb::Callback<void ()>>, 0ul, tuple<>>(void (yb::Callback<void ()>::*&)() const, tuple<yb::Callback<void ()>>&, std::__tuple_indices<0ul>, tuple<>&&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/bind.h:295:     @     0x7f6ad6480a5f  std::__bind_return<void (yb::Callback<void ()>::*)() const, tuple<yb::Callback<void ()>>, tuple<>, __is_valid_bind_return<void (yb::Callback<void ()>::*)() const, tuple<yb::Callback<void ()>>, tuple<>>::value>::type std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>::operator()[abi:v15002]<>()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:394:     @     0x7f6ad6480a34  decltype(std::declval<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>&>()()) std::__invoke[abi:v15002]<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>&>(std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:479:     @     0x7f6ad64809f4  void std::__invoke_void_return_wrapper<void, true>::__call<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>&>(std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:185:     @     0x7f6ad64809cc  std::__function::__alloc_func<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>, std::allocator<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>>, void ()>::operator()[abi:v15002]()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:359:     @     0x7f6ad647fda8  std::__function::__func<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>, std::allocator<std::__bind<void (yb::Callback<void ()>::*)() const, yb::Callback<void ()>&>>, void ()>::operator()()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:512:     @     0x7f6addbef171  std::__function::__value_func<void ()>::operator()[abi:v15002]() const
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:1187:     @     0x7f6addbbcc24  std::function<void ()>::operator()() const
../../src/yb/util/threadpool.h:510:                                                                     @     0x7f6ad6480c38  yb::FunctionRunnable::Run()
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/util/threadpool.cc:612:     @     0x7f6ad6479bae  yb::ThreadPool::DispatchThread(bool)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:359:     @     0x7f6ad648e2a8  decltype(*std::declval<yb::ThreadPool*&>().*std::declval<void (yb::ThreadPool::*&)(bool)>()(std::declval<bool&>())) std::__invoke[abi:v15002]<void (yb::ThreadPool::*&)(bool), yb::ThreadPool*&, bool&, void>(void (yb::ThreadPool::*&)(bool), yb::ThreadPool*&, bool&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/bind.h:263:     @     0x7f6ad648e229  std::__bind_return<void (yb::ThreadPool::*)(bool), tuple<yb::ThreadPool*, bool>, tuple<>, __is_valid_bind_return<void (yb::ThreadPool::*)(bool), tuple<yb::ThreadPool*, bool>, tuple<>>::value>::type std::__apply_functor[abi:v15002]<void (yb::ThreadPool::*)(bool), tuple<yb::ThreadPool*, bool>, 0ul, 1ul, tuple<>>(void (yb::ThreadPool::*&)(bool), tuple<yb::ThreadPool*, bool>&, std::__tuple_indices<0ul, 1ul>, tuple<>&&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/bind.h:295:     @     0x7f6ad648e1bf  std::__bind_return<void (yb::ThreadPool::*)(bool), tuple<yb::ThreadPool*, bool>, tuple<>, __is_valid_bind_return<void (yb::ThreadPool::*)(bool), tuple<yb::ThreadPool*, bool>, tuple<>>::value>::type std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>::operator()[abi:v15002]<>()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:394:     @     0x7f6ad648e194  decltype(std::declval<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>&>()()) std::__invoke[abi:v15002]<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>&>(std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/invoke.h:479:     @     0x7f6ad648e154  void std::__invoke_void_return_wrapper<void, true>::__call<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>&>(std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>&)
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:185:     @     0x7f6ad648e12c  std::__function::__alloc_func<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>, std::allocator<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>>, void ()>::operator()[abi:v15002]()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:359:     @     0x7f6ad648d668  std::__function::__func<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>, std::allocator<std::__bind<void (yb::ThreadPool::* const&)(bool), yb::ThreadPool* const&, bool const&>>, void ()>::operator()()
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:512:     @     0x7f6addbef171  std::__function::__value_func<void ()>::operator()[abi:v15002]() const
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221029203613-efbc29b551-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/__functional/function.h:1187:     @     0x7f6addbbcc24  std::function<void ()>::operator()() const
/home/amartsin/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/util/thread.cc:800:      @     0x7f6ad645b5a6  yb::Thread::SuperviseThread(void*)
./nptl/pthread_create.c:442:                                                                            @     0x7f6ad4c29b42  start_thread
../sysdeps/unix/sysv/linux/x86_64/clone3.S:81:                                                          @     0x7f6ad4cbb9ff  (unknown)
```

@andrei-mart from the trace above this looks like a debug build correct? Does it also fail in release?

@bmatican @rthallamko3 Still is it expected that timeout above will lead to a fatal? In what cases would that timeout happen?

**Comment by def-** [comment]
With @andrei-mart's instructions this can be reproduced perfectly. In our QA testing we have seen this FATAL timeout in different scenarios than adding foreign keys too, are all of these issues connected?
@qvad ran into it in stress: https://github.com/yugabyte/yugabyte-db/issues/11089
I ran into it in LST: https://github.com/yugabyte/yugabyte-db/issues/12926
@shamanthchandra-yb ran into it as well, looks like also stress: https://github.com/yugabyte/yugabyte-db/issues/15053

@m-iancu I assigned you so we have an assignee at all.

**Comment by bmatican** [comment]
@def- yes, this type of FATAL should be generic, around some form of DDLs waiting on backgorund operations, then crashing if not done in 60s (gflag), but we hadn't pinned down a concrete easy local repro

**Comment by bmatican** [comment]
@def- do you have the repro, as a simple 3-4 steps, with an artificial generate_series to preload the initial table? since Andrei's initial issue comment assumes an existing user table

**Comment by andrei-mart** [comment]
I don't know about the release, I'm not looking into the details.
I just tried to repeat customer's scenario before the meeting and ran into what seems like reproducible case. 
Though I'm quite sure the customer used release build.

To me, the case is failing because we have a timeout where lengthy operation should be expected. The release performs better, so it make prevent edge test case from failing, but it still would fail in larger case.

**Comment by def-** [comment]
> @def- do you have the repro, as a simple 3-4 steps, with an artificial generate_series to preload the initial table? since Andrei's initial issue comment assumes an existing user table

I just ran these commands on macOS with master yugabyte-db build and got the same FATAL as Andrei:
```
yugabyte=# create table mytest129(mykey int primary key, myvalue text);
CREATE TABLE
Time: 1163.959 ms (00:01.164)
yugabyte=# insert into mytest129 select i, 'mytest ' || i::text from generate_series(1, 2180302) i;
INSERT 0 2180302
Time: 242104.873 ms (04:02.105)
yugabyte=# create table mychild129(mykey int, myseq int, myvalue text, primary key (mykey, myseq));
CREATE TABLE
Time: 446.205 ms
yugabyte=# insert into mychild129 select i, j, 'mychild ' || i::text || ', ' || j::text  from generate_series(1, 2180302) i, generate_series(1, 2) j;
INSERT 0 4360604
Time: 608915.706 ms (10:08.916)
yugabyte=# alter table mychild129 add foreign key (mykey) references mytest129(mykey) on delete cascade;
ERROR:  Timed out waiting for Alter Table
Time: 60018.922 ms (01:00.019)
```

**Comment by bmatican** [comment]
Thanks @def- I was able to repro this easily with `/tmp/test.sql`
```
create table mytest129(mykey int primary key, myvalue text);
insert into mytest129 select i, 'mytest ' || i::text from generate_series(1, 100000) i;
create table mychild129(mykey int, myseq int, myvalue text, primary key (mykey, myseq));
insert into mychild129 select i, j, 'mychild ' || i::text || ', ' || j::text  from generate_series(1, 100000) i, generate_series(1, 2) j;
alter table mychild129 add foreign key (mykey) references mytest129(mykey) on delete cascade;
create table mytest129(mykey int primary key, myvalue text);
insert into mytest129 select i, 'mytest ' || i::text from generate_series(1, 100000) i;
create table mychild129(mykey int, myseq int, myvalue text, primary key (mykey, myseq));
insert into mychild129 select i, j, 'mychild ' || i::text || ', ' || j::text  from generate_series(1, 100000) i, generate_series(1, 2) j;
alter table mychild129 add foreign key (mykey) references mytest129(mykey) on delete cascade;
```
and
```
./bin/yb-ctl create --tserver_flags="tablet_rocksdb_ops_quiet_down_timeout_ms=5000"
 cat /tmp/test.sql | ./bin/ysqlsh
CREATE TABLE
INSERT 0 100000
CREATE TABLE
INSERT 0 200000
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
connection to server was lost
```
And then seeing the fatal in `$HOME/yugabyte-data/node-1/disk-1/yb-data/tserver/logs/yb-tserver.FATAL`

**Comment by bmatican** [comment]
Added a pg_sleep(10) before the alter and observed the following 2 threads
```
<tr><td>consensus [worker]-27353</td><td>4.770s</td><td>0.000s</td><td>0.000s</td><td rowspan="1"><pre>    @     0x7ff443e0727f·
../../src/yb/rocksdb/util/coding.h:151:                                                                 @     0x7ff4480b2cd3  rocksdb::GetVarint32Ptr(char const*, char const*, unsigned int*)
../../src/yb/rocksdb/util/coding.h:331:                                                                 @     0x7ff4480b2596  rocksdb::GetLengthPrefixedSlice(char const*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/memtable.cc:207:     @     0x7ff4480acbd4  rocksdb::MemTable::KeyComparator::operator()(char const*, char const*) const
../../src/yb/rocksdb/db/skiplist.h:345:                                                                 @     0x7ff448171594  rocksdb::SkipListBase<char const*, rocksdb::MemTableRep::KeyComparator const&, rocksdb::SingleWriterInlineSkipListNode>::KeyIsAfterNode(char const*, rocksdb::SingleWriterInlineSkipListNode*) const
../../src/yb/rocksdb/db/skiplist.h:364:                                                                 @     0x7ff448171fd1  rocksdb::SkipListBase<char const*, rocksdb::MemTableRep::KeyComparator const&, rocksdb::SingleWriterInlineSkipListNode>::FindGreaterOrEqual(char const*) const
../../src/yb/rocksdb/db/skiplist.h:314:                                                                 @     0x7ff448172433  rocksdb::SkipListBase<char const*, rocksdb::MemTableRep::KeyComparator const&, rocksdb::SingleWriterInlineSkipListNode>::Iterator::Seek(char const*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/memtable/skiplistrep.cc:139:     @     0x7ff44816bec4  rocksdb::(anonymous namespace)::SkipListRep<rocksdb::SingleWriterInlineSkipList<rocksdb::MemTableRep::KeyComparator const&>>::Iterator::Seek(yb::Slice const&, char const*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/memtable.cc:278:     @     0x7ff4480b525f  rocksdb::MemTableIterator::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/db_iter.cc:759:     @     0x7ff448062d92  rocksdb::DBIter::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/db_iter.cc:890:     @     0x7ff448066c81  rocksdb::ArenaWrappedDBIter::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/bounded_rocksdb_iterator.cc:66:     @     0x7ff44c6cadda  yb::docdb::BoundedRocksDbIterator::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/rocksdb_writer.cc:508:     @     0x7ff44c8eb88d  yb::docdb::ApplyIntentsContext::Entry(yb::Slice const&, yb::Slice const&, bool, rocksdb::DirectWriteHandler*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/rocksdb_writer.cc:429:     @     0x7ff44c8eac08  yb::docdb::IntentsWriter::Apply(rocksdb::DirectWriteHandler*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/write_batch.cc:984:     @     0x7ff44814b21d  rocksdb::DirectInsert(rocksdb::WriteBatch::Handler*, rocksdb::DirectWriter*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/write_batch.cc:342:     @     0x7ff44814a48d  rocksdb::WriteBatch::Iterate(rocksdb::WriteBatch::Handler*) const

Total number of threads: 1</pre></td></tr>
<tr><td>consensus [worker]-27350</td><td>6.270s</td><td>0.000s</td><td>0.000s</td><td rowspan="1"><pre>    @     0x7ff443e0727f·
/opt/yb-build/thirdparty/yugabyte-db-thirdparty-v20221019013514-1aec6dda08-centos7-x86_64-clang15/installed/uninstrumented/libcxx/include/c++/v1/atomic:1560:     @     0x7ff448171ae2  std::__atomic_base<rocksdb::SingleWriterInlineSkipListNode*, false>::load[abi:v15002](std::memory_order) const
../../src/yb/rocksdb/db/skiplist.h:576:                                                                 @     0x7ff448171a7e  rocksdb::SingleWriterInlineSkipListNode::Next(int)
../../src/yb/rocksdb/db/skiplist.h:360:                                                                 @     0x7ff448171e91  rocksdb::SkipListBase<char const*, rocksdb::MemTableRep::KeyComparator const&, rocksdb::SingleWriterInlineSkipListNode>::FindGreaterOrEqual(char const*) const
../../src/yb/rocksdb/db/skiplist.h:314:                                                                 @     0x7ff448172433  rocksdb::SkipListBase<char const*, rocksdb::MemTableRep::KeyComparator const&, rocksdb::SingleWriterInlineSkipListNode>::Iterator::Seek(char const*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/memtable/skiplistrep.cc:139:     @     0x7ff44816bec4  rocksdb::(anonymous namespace)::SkipListRep<rocksdb::SingleWriterInlineSkipList<rocksdb::MemTableRep::KeyComparator const&>>::Iterator::Seek(yb::Slice const&, char const*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/memtable.cc:278:     @     0x7ff4480b525f  rocksdb::MemTableIterator::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/db_iter.cc:759:     @     0x7ff448062d92  rocksdb::DBIter::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/db_iter.cc:890:     @     0x7ff448066c81  rocksdb::ArenaWrappedDBIter::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/bounded_rocksdb_iterator.cc:66:     @     0x7ff44c6cadda  yb::docdb::BoundedRocksDbIterator::Seek(yb::Slice const&)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/rocksdb_writer.cc:508:     @     0x7ff44c8eb88d  yb::docdb::ApplyIntentsContext::Entry(yb::Slice const&, yb::Slice const&, bool, rocksdb::DirectWriteHandler*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/docdb/rocksdb_writer.cc:429:     @     0x7ff44c8eac08  yb::docdb::IntentsWriter::Apply(rocksdb::DirectWriteHandler*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/write_batch.cc:984:     @     0x7ff44814b21d  rocksdb::DirectInsert(rocksdb::WriteBatch::Handler*, rocksdb::DirectWriter*)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/write_batch.cc:342:     @     0x7ff44814a48d  rocksdb::WriteBatch::Iterate(rocksdb::WriteBatch::Handler*) const
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/write_batch.cc:928:     @     0x7ff44814c28f  rocksdb::WriteBatchInternal::InsertInto(rocksdb::autovector<rocksdb::WriteThread::Writer*, 8ul> const&, unsigned long, rocksdb::ColumnFamilyMemTables*, rocksdb::FlushScheduler*, bool, unsigned long, rocksdb::DB*, yb::EnumBitSet<rocksdb::InsertFlag>)
/n/users/bogdan/code/yugabyte-db/build/debug-clang15-dynamic-ninja/../../src/yb/rocksdb/db/db_impl.cc:5266:     @     0x7ff447fe9bff  rocksdb::DBImpl::WriteImpl(rocksdb::WriteOptions const&, rocksdb::WriteBatch*, rocksdb::WriteCallback*)

Total number of threads: 1</pre></td></tr>
```
Just need to figure out if bg applies also take a ScopedRWOperation cc @robertsami

**Comment by robertsami** [comment]
using https://gist.github.com/robertsami/45c8c4b80662840c8411976d07181cdf

got the following output:
```
E1213 11:50:45.993973 99303424 operation_counter.cc:179] Hit deadline
E1213 11:50:45.994261 99303424 operation_counter.cc:182] Source MaxPersistentOpId, count: 0
E1213 11:50:45.994290 99303424 operation_counter.cc:182] Source ProcessApply, count: 1
E1213 11:50:45.994311 99303424 operation_counter.cc:182] Source StillHasOrphaned, count: 0
E1213 11:50:45.994331 99303424 operation_counter.cc:182] Source GetRegulardDbStat, count: 0
E1213 11:50:45.994351 99303424 operation_counter.cc:182] Source DoCleanupIntentFiles, count: 0
E1213 11:50:45.994370 99303424 operation_counter.cc:182] Source HasSSTables, count: 0
E1213 11:50:45.994390 99303424 operation_counter.cc:182] Source EnableCompactions, count: 0
E1213 11:50:45.994410 99303424 operation_counter.cc:182] Source WriteQueryInitExecute, count: 0
E1213 11:50:45.994428 99303424 operation_counter.cc:182] Source CleanupIntentFiles, count: 0
E1213 11:50:45.994448 99303424 operation_counter.cc:182] Source FlushIfNecessary, count: 0
E1213 11:50:45.994467 99303424 operation_counter.cc:182] Source loader, count: 0
E1213 11:50:45.994488 99303424 operation_counter.cc:184] End debug out
```

this implicates the `ProcessApply` call as responsible for blocking the alter table operation
```
  Status ProcessApply(const TransactionApplyData& data) {
    VLOG_WITH_PREFIX(2) << "Apply: " << data.ToString();

    loader_.WaitLoaded(data.transaction_id);

    ScopedRWOperation operation("ProcessApply", pending_op_counter_);
```

**Comment by m-iancu** [comment]
Possibly duplicate of https://github.com/yugabyte/yugabyte-db/issues/12926

**Comment by rthallamko3** [comment]
Based on the offline discussions, the above issue can be resolved by the following - Remove that protection of PauseReadWriteOperations, since it seems alter schema modifies only the table info and any concurrent readers would access a snapshot of that table info anyways. cc @yusong-yan

**Comment by rthallamko3** [comment]
Yusong has a review out for this. We are currently validating the fix.
